File: xform-go/xmlmodel.go
package xform

import (
	"encoding/xml"
	"io"
	"sort"
	"strings"
)

type Node struct {
	Kind      string
	Name      string
	Value     string
	Children  []*Node
	Attrs     map[string]string
	AttrOrder []string
	Parent    *Node
}

func (n *Node) StringValue() string {
	switch n.Kind {
	case "text", "attribute":
		return n.Value
	case "element", "document":
		out := ""
		for _, c := range n.Children {
			out += c.StringValue()
		}
		return out
	default:
		return ""
	}
}

func ParseXML(text string) (*Node, error) {
	return ParseXMLBytes([]byte(text))
}

func ParseXMLBytes(data []byte) (*Node, error) {
	text := normalizeXMLBytes(data)
	doc := &Node{Kind: "document", Attrs: map[string]string{}}
	decoder := xml.NewDecoder(strings.NewReader(text))
	var stack []*Node
	for {
		tok, err := decoder.Token()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, err
		}
		switch t := tok.(type) {
		case xml.StartElement:
			order := make([]string, 0, len(t.Attr))
			n := &Node{Kind: "element", Name: t.Name.Local, Attrs: map[string]string{}}
			for _, a := range t.Attr {
				n.Attrs[a.Name.Local] = a.Value
				order = append(order, a.Name.Local)
			}
			n.AttrOrder = order
			if len(stack) == 0 {
				n.Parent = doc
				doc.Children = append(doc.Children, n)
			} else {
				parent := stack[len(stack)-1]
				n.Parent = parent
				parent.Children = append(parent.Children, n)
			}
			stack = append(stack, n)
		case xml.EndElement:
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		case xml.CharData:
			if len(stack) == 0 {
				continue
			}
			txt := string(t)
			n := &Node{Kind: "text", Value: txt, Attrs: map[string]string{}}
			parent := stack[len(stack)-1]
			n.Parent = parent
			parent.Children = append(parent.Children, n)
		case xml.Comment:
			if len(stack) == 0 {
				continue
			}
			n := &Node{Kind: "comment", Value: string(t), Attrs: map[string]string{}}
			parent := stack[len(stack)-1]
			n.Parent = parent
			parent.Children = append(parent.Children, n)
		case xml.ProcInst:
			if len(stack) == 0 {
				continue
			}
			n := &Node{Kind: "pi", Value: string(t.Inst), Attrs: map[string]string{}}
			parent := stack[len(stack)-1]
			n.Parent = parent
			parent.Children = append(parent.Children, n)
		}
	}
	return doc, nil
}

func DeepCopy(node *Node, recurse bool) *Node {
	copied := &Node{Kind: node.Kind, Name: node.Name, Value: node.Value, Attrs: map[string]string{}, AttrOrder: append([]string{}, node.AttrOrder...)}
	for k, v := range node.Attrs {
		copied.Attrs[k] = v
	}
	if recurse {
		for _, c := range node.Children {
			child := DeepCopy(c, true)
			child.Parent = copied
			copied.Children = append(copied.Children, child)
		}
	}
	return copied
}

func IterDescendants(node *Node) []*Node {
	out := []*Node{}
	for _, child := range node.Children {
		out = append(out, child)
		out = append(out, IterDescendants(child)...)
	}
	return out
}

func Serialize(item *Node) string {
	switch item.Kind {
	case "document":
		out := ""
		for _, c := range item.Children {
			out += Serialize(c)
		}
		return out
	case "text":
		return escapeText(item.Value)
	case "attribute":
		return escapeAttr(item.Value)
	case "element":
		attrs := ""
		keys := item.AttrOrder
		if len(keys) == 0 {
			keys = make([]string, 0, len(item.Attrs))
			for k := range item.Attrs {
				keys = append(keys, k)
			}
			sort.Strings(keys)
		}
		for _, k := range keys {
			attrs += " " + k + "=\"" + escapeAttr(item.Attrs[k]) + "\""
		}
		if len(item.Children) == 0 {
			return "<" + item.Name + attrs + "/>"
		}
		inner := ""
		for _, c := range item.Children {
			inner += Serialize(c)
		}
		return "<" + item.Name + attrs + ">" + inner + "</" + item.Name + ">"
	default:
		return ""
	}
}

func escapeText(text string) string {
	replacer := strings.NewReplacer("&", "&amp;", "<", "&lt;", ">", "&gt;")
	return replacer.Replace(text)
}

func escapeAttr(text string) string {
	replacer := strings.NewReplacer("\"", "&quot;")
	return replacer.Replace(escapeText(text))
}

func normalizeXMLBytes(data []byte) string {
	text := string(data)
	lower := strings.ToLower(text)
	if strings.Contains(lower, "encoding=\"iso-8859-1\"") || strings.Contains(lower, "encoding='iso-8859-1'") {
		runes := make([]rune, 0, len(data))
		for _, b := range data {
			runes = append(runes, rune(b))
		}
		text = string(runes)
		text = strings.ReplaceAll(text, "encoding=\"ISO-8859-1\"", "encoding=\"UTF-8\"")
		text = strings.ReplaceAll(text, "encoding='ISO-8859-1'", "encoding=\"UTF-8\"")
	}
	return replaceNamedEntities(text)
}

func replaceNamedEntities(text string) string {
	replacer := strings.NewReplacer(
		"&mdash;", "—",
		"&hellip;", "…",
		"&nbsp;", "\u00a0",
	)
	return replacer.Replace(text)
}

File: xform-kotlin/src/main/kotlin/com/zopyx/xform/XmlModel.kt
package com.zopyx.xform

import org.w3c.dom.Document
import org.w3c.dom.Node as DomNode
import java.io.StringReader
import javax.xml.parsers.DocumentBuilderFactory
import org.xml.sax.InputSource

class XmlNode(
    val kind: String,
    var name: String = "",
    var value: String = "",
    val children: MutableList<XmlNode> = mutableListOf(),
    val attrs: MutableMap<String, String> = mutableMapOf(),
    var attrOrder: MutableList<String> = mutableListOf(),
    var parent: XmlNode? = null
) {
    fun stringValue(): String {
        return when (kind) {
            "text", "attribute" -> value
            "element", "document" -> children.joinToString("") { it.stringValue() }
            else -> ""
        }
    }
}

object XmlModel {
    fun parseXMLBytes(data: ByteArray): XmlNode {
        val text = normalizeXMLBytes(data)
        val factory = DocumentBuilderFactory.newInstance().apply {
            isNamespaceAware = false
            isExpandEntityReferences = true
            setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false)
            setFeature("http://xml.org/sax/features/external-general-entities", false)
            setFeature("http://xml.org/sax/features/external-parameter-entities", false)
        }
        val builder = factory.newDocumentBuilder()
        builder.setEntityResolver { _, _ -> InputSource(StringReader("")) }
        val doc = builder.parse(InputSource(StringReader(text)))
        val root = XmlNode("document")
        convertChildren(doc, root)
        return root
    }

    private fun convertChildren(domParent: DomNode, outParent: XmlNode) {
        val list = domParent.childNodes
        for (i in 0 until list.length) {
            val c = list.item(i)
            when (c.nodeType) {
                DomNode.ELEMENT_NODE -> {
                    val n = XmlNode("element", name = c.nodeName)
                    val attrs = c.attributes
                    for (j in 0 until attrs.length) {
                        val a = attrs.item(j) as org.w3c.dom.Attr
                        n.attrs[a.name] = a.value
                        n.attrOrder.add(a.name)
                    }
                    n.parent = outParent
                    outParent.children.add(n)
                    convertChildren(c, n)
                }
                DomNode.TEXT_NODE, DomNode.CDATA_SECTION_NODE -> {
                    val n = XmlNode("text", value = c.nodeValue)
                    n.parent = outParent
                    outParent.children.add(n)
                }
                DomNode.COMMENT_NODE -> {
                    val n = XmlNode("comment", value = c.nodeValue)
                    n.parent = outParent
                    outParent.children.add(n)
                }
                DomNode.PROCESSING_INSTRUCTION_NODE -> {
                    val n = XmlNode("pi", value = (c as org.w3c.dom.ProcessingInstruction).data)
                    n.parent = outParent
                    outParent.children.add(n)
                }
            }
        }
    }

    fun deepCopy(node: XmlNode, recurse: Boolean = true): XmlNode {
        val copied = XmlNode(
            kind = node.kind,
            name = node.name,
            value = node.value,
            attrs = node.attrs.toMutableMap(),
            attrOrder = node.attrOrder.toMutableList()
        )
        if (recurse) {
            for (c in node.children) {
                val child = deepCopy(c, true)
                child.parent = copied
                copied.children.add(child)
            }
        }
        return copied
    }

    fun iterDescendants(node: XmlNode): List<XmlNode> {
        val out = mutableListOf<XmlNode>()
        for (child in node.children) {
            out.add(child)
            out.addAll(iterDescendants(child))
        }
        return out
    }

    fun serialize(node: XmlNode): String {
        return when (node.kind) {
            "document" -> node.children.joinToString("") { serialize(it) }
            "text" -> escapeText(node.value)
            "attribute" -> escapeAttr(node.value)
            "element" -> {
                val keys = if (node.attrOrder.isEmpty()) {
                    node.attrs.keys.sorted()
                } else {
                    node.attrOrder
                }
                val attrs = keys.joinToString("") { " $it=\"${escapeAttr(node.attrs[it] ?: "")}\"" }
                if (node.children.isEmpty()) {
                    "<${node.name}$attrs/>"
                } else {
                    val inner = node.children.joinToString("") { serialize(it) }
                    "<${node.name}$attrs>$inner</${node.name}>"
                }
            }
            else -> ""
        }
    }

    private fun escapeText(text: String): String {
        return text.replace("&", "&amp;")
            .replace("<", "&lt;")
            .replace(">", "&gt;")
    }

    private fun escapeAttr(text: String): String {
        return escapeText(text).replace("\"", "&quot;")
    }

    private fun normalizeXMLBytes(data: ByteArray): String {
        var text = data.toString(Charsets.UTF_8)
        val lower = text.lowercase()
        if ("encoding=\"iso-8859-1\"" in lower || "encoding='iso-8859-1'" in lower) {
            text = data.toString(Charsets.ISO_8859_1)
                .replace("encoding=\"ISO-8859-1\"", "encoding=\"UTF-8\"")
                .replace("encoding='ISO-8859-1'", "encoding=\"UTF-8\"")
        }
        return replaceNamedEntities(text)
    }

    private fun replaceNamedEntities(text: String): String {
        return text.replace("&mdash;", "—")
            .replace("&hellip;", "…")
            .replace("&nbsp;", "\u00A0")
    }
}
